src/ray/core_worker/core_worker.cc

```diff
@@ -480,18 +480,23 @@ void CoreWorker::Shutdown() {
     options_.on_worker_shutdown(GetWorkerID());
   }
 
-  io_service_.stop();
   if (gcs_client_) {
-    // TODO(sang): This causes a thread-safety bug because
-    // it invalidates a pointer that's used by a different thread.
-    // We should run Disconnect in a io_service thread.
+    // We should disconnect gcs client first otherwise because it contains
+    // a blocking logic that can block the io service upon
+    // gcs shutdown.
+    // TODO(sang): Refactor GCS client to be more robust.
     RAY_LOG(INFO) << "Disconnecting a GCS client.";
     gcs_client_->Disconnect();
```

Comment:
Note that Disconnect before io_service.join is necessary because io service has a blocking logic (we should remove it) which can block io service thread when gcs shuts down. If that happens without calling Disconnect, the worker will QuickExit